### PR TITLE
Refactoring of prepdocs for easier integration with user upload feature

### PR DESCRIPTION
**Pull request by pamelafox**
## Purpose

This PR includes only the changes to prepdocs that are needed for the user document upload feature in #1395. This way, we can merge these changes first, and then that PR will simply *move* the files into app/backend, instead of move and edit.

The notable changes are:

* print -> logger.info(). We use our own logger so that we can set its level to INFO. If we set global level to INFO, then we also get Azure Python SDK stuff that's quite noisy right now.
* Reorg of the functions in prepdocs.py to be more reusable by app/backend, with additional typing and validation. No more passing around of args!
* Additional option to remove_content to remove documents with a single oid assigned
* Changed the id function to include an acl hash if acls are defined. The goal is to avoid id collisions between user-uploaded files and company-uploaded files.

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[X] No - Tested in a manual ingestion env and int vect env.
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[ ] Yes
[X] No
```

## Type of change

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[X] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](../CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [X] I added tests that prove my fix is effective or that my feature works: **Added tests for everything but filestrategy.py as that currently has no tests, requires a ton of mocking.**
- [X] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [X] I ran `python -m mypy` to check for type errors
- [X] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Comment by pamelafox** [comment]
I think it'd be good to merge https://github.com/Azure-Samples/azure-search-openai-demo/pull/1355 before this one as they both touch same lines.

File: scripts/prepdocs.py [hunk 1]
```diff
@@ -32,50 +32,159 @@
...
+async def setup_search_info(
+    search_service: str,
+    index_name: str,
+    azure_credential: AsyncTokenCredential,
+    search_key: Union[str, None] = None,
+    key_vault_name: Union[str, None] = None,
+    search_secret_name: Union[str, None] = None,
+) -> SearchInfo:
+    if key_vault_name and search_secret_name:
+        key_vault_client = SecretClient(
```
Comment (mattgotteiner) [inline scripts/prepdocs.py:54]: does this support async with?

File: scripts/prepdocs.py [hunk 2]
```diff
@@ -388,28 +364,24 @@ async def main(strategy: Strategy, credential: AsyncTokenCredential, args: Any):
...
     parser.add_argument(
         "--useintvectorization",
         required=False,
         help="Required if --useintvectorization is specified. Enable Integrated vectorizer indexer support which is in preview)",
     )
     parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")
     args = parser.parse_args()
+
+    if args.verbose:
+        # set formatter to only show the message
```
Comment (mattgotteiner) [inline scripts/prepdocs.py:381]: should we comment to state this is for the SDK workaround as discussed?

**Review by mattgotteiner — APPROVED** [review]

File: scripts/prepdocs.py (hunk 1 above)
Comment (pamelafox) [reply scripts/prepdocs.py:54]: yep! changing to that.

File: scripts/prepdocs.py (hunk 2 above)
Comment (pamelafox) [reply scripts/prepdocs.py:381]: Good idea, adding

**Comment by javapro13** [comment]
Hello @pamelafox I just installed the latest from https://github.com/Azure-Samples/azure-search-openai-demo/tree/main and do not see the + manage file upload button on the main page. Do I need to do anything additional for the users to be able to upload the files? This is a critical functionality for us

**Comment by pamelafox** [comment]
@javapro13 Yes, you need to follow the steps here to enable it: https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/docs/deploy_features.md#enabling-user-document-upload

**Pull request merged**